File: Prime Numbers/Range Of Prime/Sieve Of Eratosthenes/Method3/SOE3.c
#include <stdio.h>
#include <stdlib.h>

// Sieve Function
char *seive(int max) {

	// Reviesed Max
	int seive_limit = max/2;
	// Set index to first not prime no. index
	int initial = 4;

	// Prime list
	char *p = (char *)malloc((seive_limit+1)*sizeof(char));

	// Set Value to true
	for (int i=0; i<=seive_limit; i++) { p[i] = '1'; }

	// Marking marking not prime no.
	for (int step=3; step<=max; step+=2) {
		for (int notPrime_index=initial;
			notPrime_index<seive_limit;
			notPrime_index+=step)
			{ p[notPrime_index-1] = '0'; }
		initial += 2*(step+1);

		// Once, all index are checked exit.
		if (initial > seive_limit) { return p; }
	}

}

// DRIVER CODE
void main() {
	// Upmost limit
	int max;
	scanf("%d", &max);

	// Get Primelist
	char *primeList = seive( max );

	// Print prime list
	printf("2 ");
	for ( int i=0; i<(max-1)/2; i++ ) {
		// Print on if i is checked
		if ( primeList[i]=='1' ) {
			// i is odd index of no. N and i start from 0 so N will be 2*(i+1)+1
			printf("%d ", 2*(i+1)+1);
		}
	}
	printf("\n");
}

File: Prime Numbers/Range Of Prime/Sieve Of Eratosthenes/Method3/SOE3.py
# Sieve Function
def seive(max):
	# Make prime list
	p = list( range(3, max+1, 2) )
	# Revised Max limit
	seive_limit = len(p)
	# Set initial index for notprime no.
	initial = 4

	# Iterate through list and mark not prime no.
	for step in range( 3, max+1, 2 ):
		for i in range( initial, seive_limit+1, step ):
			p[i-1]=0
		initial += 2*(step+1)

		# If all values are checked
		if initial > seive_limit:
			return [2] + list(filter(None, p))


if __name__ == "__main__":
	max = int(input())		# Max limit
	primeList = seive(max)	# Get prime list
	print(primeList);		# Print list